File: examples/recognize-demo-java/src/test/java/com/provekit/demo/recognize/PersistTest.java
package com.provekit.demo.recognize;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import org.junit.jupiter.api.Test;

class PersistTest {
    @Test
    void openConnectionOpensSqliteJdbcUrl() throws Exception {
        try (Connection conn = Persist.openConnection("jdbc:sqlite::memory:")) {
            assertFalse(conn.isClosed());
        }
    }

    @Test
    void executeStatementRunsDdlAndDml() throws Exception {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:")) {
            try (PreparedStatement create = conn.prepareStatement(
                    "CREATE TABLE events (id INTEGER PRIMARY KEY, user TEXT NOT NULL)")) {
                Persist.executeStatement(create);
            }

            int changed;
            try (PreparedStatement insert = conn.prepareStatement("INSERT INTO events (user) VALUES ('alice')")) {
                changed = Persist.executeStatement(insert);
            }

            assertEquals(1, changed);
        }
    }

    @Test
    void queryRowsReturnsCursorForStatementResults() throws Exception {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:")) {
            try (PreparedStatement create = conn.prepareStatement(
                    "CREATE TABLE events (id INTEGER PRIMARY KEY, user TEXT NOT NULL)")) {
                Persist.executeStatement(create);
            }
            try (PreparedStatement insert = conn.prepareStatement("INSERT INTO events (user) VALUES ('alice')")) {
                Persist.executeStatement(insert);
            }

            try (ResultSet rows = Persist.queryRows(conn, "SELECT user FROM events")) {
                assertTrue(rows.next());
                assertEquals("alice", rows.getString(1));
                assertFalse(rows.next());
            }
        }
    }
}

File: examples/recognize-demo-go/pkg/persist/persist_test.go
package persist

import "testing"

func TestOpenStoreReturnsUsableInMemorySQLite(t *testing.T) {
	db := OpenStore("sqlite", ":memory:")
	defer db.Close()

	if _, err := db.Exec(`create table events (body text not null)`); err != nil {
		t.Fatalf("create table: %v", err)
	}
	if _, err := db.Exec(`insert into events (body) values (?)`, `{"ok":true}`); err != nil {
		t.Fatalf("insert event: %v", err)
	}

	var body string
	if err := db.QueryRow(`select body from events`).Scan(&body); err != nil {
		t.Fatalf("select body: %v", err)
	}
	if body != `{"ok":true}` {
		t.Fatalf("body = %q", body)
	}
}
